File: src/backend/wireless.py
import sys
import subprocess


def get_wireless_status():
    status = {
        "wifi_enabled": False,
        "wifi_connected": False,
        "connection_type": None,  # wifi / ethernet / none
        "bluetooth_enabled": False,
        "bluetooth_connected": False
    }

    if sys.platform.startswith("linux"):

        # ---------------------------
        # 🌐 REDE (WIFI OU CABO)
        # ---------------------------
        try:
            output = subprocess.check_output(
                ["nmcli", "-t", "-f", "TYPE,STATE", "device"],
                text=True
            )

            for line in output.splitlines():
                dev_type, state = line.split(":")

                if state == "connected":
                    status["wifi_connected"] = True
                    status["wifi_enabled"] = True
                    status["connection_type"] = dev_type  # wifi / ethernet
                    break

        except Exception:
            pass

        # ---------------------------
        # 📶 WIFI (estado geral)
        # ---------------------------
        try:
            wifi_state = subprocess.check_output(
                ["nmcli", "radio", "wifi"],
                text=True
            ).strip()

            status["wifi_enabled"] = wifi_state.lower() == "enabled"

        except Exception:
            pass

        # ---------------------------
        # 🔵 BLUETOOTH
        # ---------------------------
        try:
            bt_state = subprocess.check_output(
                ["bluetoothctl", "show"],
                text=True
            )

            status["bluetooth_enabled"] = "Powered: yes" in bt_state

            bt_conn = subprocess.check_output(
                ["bluetoothctl", "info"],
                text=True
            )

            status["bluetooth_connected"] = "Connected: yes" in bt_conn

        except Exception:
            pass

    else:
        # 💻 Ambiente de desenvolvimento (Windows, etc)
        status["wifi_enabled"] = True
        status["wifi_connected"] = True
        status["connection_type"] = "wifi"
        status["bluetooth_enabled"] = False
        status["bluetooth_connected"] = False

    return status

File: src/server/src/wireless.ts
import { execSync } from 'child_process'
import * as os from 'os'
import type { Server } from 'socket.io'

const WIRELESS_POLL_INTERVAL_MS = 5000

interface WirelessStatus {
  wifiEnabled: boolean
  wifiConnected: boolean
  connectionType: string | null // "wifi" | "ethernet" | null
  bluetoothEnabled: boolean
  bluetoothConnected: boolean
}

interface WifiNetwork {
  ssid: string
  signal: number
  security: string
  connected: boolean
}

function runCommand(cmd: string): string {
  return execSync(cmd, { encoding: 'utf-8' })
}

export function getWirelessStatus(): WirelessStatus {
  const status: WirelessStatus = {
    wifiEnabled: false,
    wifiConnected: false,
    connectionType: null,
    bluetoothEnabled: false,
    bluetoothConnected: false,
  }

  if (os.platform() === 'linux') {
    // ---------------------------
    // NETWORK (WIFI OR CABLE)
    // ---------------------------
    try {
      const output = runCommand('nmcli -t -f TYPE,STATE device')

      for (const line of output.split('\n')) {
        if (!line.trim()) continue
        const [devType, state] = line.split(':')

        if (state === 'connected') {
          status.wifiConnected = true
          status.wifiEnabled = true
          status.connectionType = devType // wifi / ethernet
          break
        }
      }
    } catch {}

    // ---------------------------
    // WIFI
    // ---------------------------
    try {
      const wifiState = runCommand('nmcli radio wifi').trim()
      status.wifiEnabled = wifiState.toLowerCase() === 'enabled'
    } catch {}

    // ---------------------------
    // BLUETOOTH
    // ---------------------------
    try {
      const btState = runCommand('bluetoothctl show')
      status.bluetoothEnabled = btState.includes('Powered: yes')

      const btConn = runCommand('bluetoothctl info')
      status.bluetoothConnected = btConn.includes('Connected: yes')
    } catch {}
  } else {
    status.wifiEnabled = true
    status.wifiConnected = true
    status.connectionType = 'ethernet'
    status.bluetoothEnabled = false
    status.bluetoothConnected = false
  }
  return status
}

export function scanWifiNetworks(): WifiNetwork[] {
  let networks: WifiNetwork[] = []

  if (os.platform() !== 'linux') {
    return networks
  }

  try {
    execSync('nmcli device wifi rescan', { stdio: 'ignore' })

    const output = runCommand(
      'nmcli -t -f IN-USE,SSID,SIGNAL,SECURITY device wifi list',
    )

    for (const line of output.split('\n')) {
      if (!line.trim()) continue
      const parts = line.split(':')

      if (parts.length >= 4) {
        const [inUse, ssid, signal, security] = parts

        networks.push({
          ssid,
          signal: /^\d+$/.test(signal) ? parseInt(signal, 10) : 0,
          security,
          connected: inUse === '*',
        })
      }
    }

    networks = networks.sort((a, b) => {
      if (a.connected !== b.connected) return a.connected ? -1 : 1
      return b.signal - a.signal
    })
  } catch (e) {
    console.error('Error scanning Wi-Fi:', e)
  }

  return networks
}

export function registerWirelessHandlers(io: Server): () => void {
  let lastStatusJson: string | null = null

  function pollAndBroadcast() {
    try {
      const status = getWirelessStatus()
      const statusJson = JSON.stringify(status)

      if (statusJson !== lastStatusJson) {
        lastStatusJson = statusJson
        io.emit('wireless:status', status)
      }
    } catch (error) {
      console.error('[Wireless] Failed to poll status:', error)
    }
  }

  pollAndBroadcast()
  const intervalHandle = setInterval(
    pollAndBroadcast,
    WIRELESS_POLL_INTERVAL_MS,
  )

  io.on('connection', (socket) => {
    if (lastStatusJson) {
      socket.emit('wireless:status', JSON.parse(lastStatusJson))
    }
  })

  return () => clearInterval(intervalHandle)
}
